Improve EnumInteger-System

JaniruTEC opened the pull request:
Added new package conv (conversion)
Added convenience methods
Added missing static methods
_Added IntegerConvertible as base interface of EnumInteger_
Added MaskValue as special case of EnumInteger that is used to express masks
Renamed EnumIntegerSet to MaskValueSet and changed it to only accept MaskValues
Made EnumIntegerSet an interface and added EnumIntegerSetImpl as implementation 
Renamed enumSetFromInt to ~getSetFromInt~ _maskValueSet_

JaniruTEC committed:
Moved EnumInteger to new package conv (conversion)

Moved EnumInteger to new package conv (conversion) as first step to refactoring the EnumInteger-System
Moved the appropiate Test
Updated imports in all accessors

JaniruTEC committed:
Moved EnumIntegerSet to package conv

Moved EnumIntegerSet to package conv
Moved the appropiate Test
Updated imports in all accessors

JaniruTEC committed:
Generified EnumInteger and added overloaded method

Generified EnumInteger to not necessarily represent a bitmask, but an intValue that represents the Enum
Modified depended classes to reflect this change
Modified Tests to reflect this change

Added overloaded version of EnumInteger#enumFromInt that accepts an enum-type instead of an array of possible values

JaniruTEC committed:
Added MaskValueEnum as interface for EnumIntegers that can be put into an EnumIntegerSet

JaniruTEC committed:
Changed appropriate classes to use MaskValueEnum

JaniruTEC committed:
Renamed EnumIntegerSet to MaskValueSet

Renamed EnumIntegerSet to MaskValueSet
Updated depended classes
Renamed appropriate Test

JaniruTEC committed:
Changed MaskValueSet to only accept MaskValueEnums

Changed MaskValueSet to only accept MaskValueEnums (and therefore match it's contract)

JaniruTEC committed:
Changed MaskValueSet to an interface and added class MaskValueSetImpl

Changed MaskValueSet to an interface and added class MaskValueSetImpl as implementor

JaniruTEC committed:
Added convenience methods to all MaskValueEnums, added overloaded method and renamed #fromInt

Renamed MaskValueSet#enumSetFromInt to #getSetFromInt
Renamed #fromInt in MaskValueEnums to #getSetFromInt
Added #getSetFromInt to MaskValueEnums where it was missing
Updated depended classes

Added overloaded version of MaskValueSet#getSetFromInt that accepts an enum-type instead of an array of possible values

Added delegate-method (getSetFromInt) to MaskValueEnum

JaniruTEC committed:
Added convenience methods to all EnumIntegers

Added #fromInt to EnumIntegers where it was missing

JaniruTEC committed:
Removed final keyword from MaskValueSetImpl

JaniruTEC committed:
Added additional delegates to MaskValueEnum

JaniruTEC committed:
Merge remote-tracking branch 'upstream/develop' into improved-enumintegers

infeo commented on src/main/java/dev/dokan/dokan_java/constants/dokany/MountError.java:
```diff
@@ -17,20 +17,21 @@
     VERSION_ERROR(-7, "Mount failed: Requested an incompatible version.");
 
-    private final int mask;
+    private final int intValue;
```
That is a good change. Since this is no mask, it shouldn't be named so.

infeo commented on src/main/java/dev/dokan/dokan_java/constants/dokany/MountOption.java:
```diff
@@ -20,20 +20,21 @@
     FILELOCK_USER_MODE(256, "Enable Lockfile/Unlockfile operations. Otherwise Dokan will take care of it.");
 
-    private final int mask;
+    private final int maskValue;
```
since this is the enum (and not the set), maybe this should also be called `intValue`. From my perspective, if we have a set, its a mask, otherwise just an int value representing a single setting.

JaniruTEC replied:
The origin of the name was not so much the idea that the value holds a mask, but is used in a mask. MaskValue is the value, which is applied to the mask.

infeo replied:
Ahh, from this perspective it makes more sense^^ Maybe it can be stressed out in a way, that it represents exactly one single value? maskValue sounds like "value of the mask". Right now, i don't have a good suggestion.

But this won't be something, which blocks the PR.

JaniruTEC replied:
@infeo and I agreed to replace `maskValue` with `maskingValue` in those cases.
Resolved with commit 8ef36a92aeafb39610c0e5e4b2872e96d50a4f2f

infeo commented on src/main/java/dev/dokan/dokan_java/constants/dokany/MountOption.java:
```diff
@@ -20,20 +20,21 @@
 
-    public static EnumIntegerSet<MountOption> fromInt(final int value) {
-        return EnumIntegerSet.enumSetFromInt(value, values());
+    public static MaskValueSet<MountOption> getSetFromInt(final int value) {
```
here the `value` could be called mask.

I suggest to change the naming of the function. Looking for example at the [Collections](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/Collections.html) class of the JDK, if you create a collection, it is not preceeded by "get" but states directly what is created. Maybe something like `maskValueSet` ?

JaniruTEC replied:
I will change `value` to `mask`. This originated solely in the old method name.

I wasn't happy with the naming myself. I will change it to your suggestion.

JaniruTEC replied:
Resolved with commit e042b3d44d5b944f198f2cd3a2dca89419f56e8f / 3114b10f633ea2c173307afff6119c208ba7d791

infeo commented on src/main/java/dev/dokan/dokan_java/conv/MaskValueEnum.java:
```diff
@@ -0,0 +1,21 @@
+
+public interface MaskValueEnum extends EnumInteger {
+
+    static <T extends Enum<T> & MaskValueEnum> T enumFromInt(final int value, final T[] enumValues) {
```
the naming of the functions needs to be discussed (see `MountOptions.java`)

JaniruTEC replied:
I presume You meant to write `MountOption.java`.

Also what would be your suggestion for `enumFromInt`?

JaniruTEC replied:
Resolved with commit e042b3d44d5b944f198f2cd3a2dca89419f56e8f / 3114b10f633ea2c173307afff6119c208ba7d791

infeo commented on src/main/java/dev/dokan/dokan_java/conv/MaskValueEnum.java:
```diff
@@ -0,0 +1,21 @@
+        return MaskValueSet.getSetFromInt(intValue, type);
+    }
+
+    default int getMaskValue() {
```
Also here, the enum is singular, therefore this mask should maybe renamed to intValue

JaniruTEC replied:
Resolved with commit e042b3d44d5b944f198f2cd3a2dca89419f56e8f / 3114b10f633ea2c173307afff6119c208ba7d791

infeo commented on src/main/java/dev/dokan/dokan_java/conv/MaskValueSet.java:
```diff
@@ -0,0 +1,56 @@
+        return elements;
+    }
+
+    static <T extends Enum<T> & MaskValueEnum> MaskValueSet<T> getSetFromInt(final int intValue, final Class<T> type) {
```
also here one should think about the naming

JaniruTEC replied:
Resolved with commit e042b3d44d5b944f198f2cd3a2dca89419f56e8f / 3114b10f633ea2c173307afff6119c208ba7d791

infeo reviewed (changes requested):
First of all. thank you for this pull request. It adds typesafety and reduces wrong use of the system. 

A thing which should be discussed is the naming of some objects. I don't like the new package name. I assume it stands for conversion, but conversion would only contain classes for the conversion action, not the conversion results. I suggest the package name `wrapper`, because the int value is wrapped in a type safe manner and in the end these objects still represent the value, therefore wrap it.

The comments for `MountOption.java` hold basically for all classes implementing the `MaskValueEnum`.

Splitting the `MaskValueEnum` into interface and implementation is overengineering for this library, but acceptable from my point of view.

JaniruTEC commented:
> A thing which should be discussed is the naming of some objects. I don't like the new package name. I assume it stands for conversion, but conversion would only contain classes for the conversion action, not the conversion results. I suggest the package name `wrapper`, because the int value is wrapped in a type safe manner and in the end these objects still represent the value, therefore wrap it.

`conv` does indeed stand for conversion. While I understand your suggestion `wrapper` wouldn't be a good name either because it's already taken by the #32-API.

> The comments for `MountOption.java` hold basically for all classes implementing the `MaskValueEnum`.

Same goes for the cases where my comments are missing.

> Splitting the `MaskValueEnum` into interface and implementation is overengineering for this library, but acceptable from my point of view.

For documentation purposes: The idea was to stand in line with the JDK-Collections, where you usually also deal with a lot of interfaces. It also makes swapping out the backend easier in case it's needed.

infeo commented on src/main/java/dev/dokan/dokan_java/conv/MaskValueEnum.java:
```diff
@@ -0,0 +1,21 @@
+        return MaskValueSet.getSetFromInt(intValue, allEnumValues);
+    }
+
+    static <T extends Enum<T> & MaskValueEnum> MaskValueSet<T> getSetFromInt(final int intValue, final Class<T> type) {
```
I'm asking myself currently, if these methods should or shouldn't be part of the MasValueEnum class.

They couple the MaskValueSet, which relies on this interface, to the interface itself and reverse the dependency.
If it is implemented in the base class, it is okay, because the type is directly fixed and inreases type safety, but the base interface should not contain any reference to MaskValueSet

JaniruTEC replied:
I was thinking about that a lot, too.
I decided to throw it in because `MaskValueEnum` and `MaskValueSet` have cyclic dependency on each other. MaskValueSet is designed to specifically only hold MaskValueEnums and MaskValueEnum is designed to be used with MaskValueSet. 

Also putting the `maskValueSet` method into MaskValueEnum creates a contract between it and all implementing classes showing where the same method "comes from" in, e.g. MountOption. 
At least that was my line of thinking here. Java sadly doesn't support static inheritance.

infeo replied:
I vote against it.
An alternative can be having an inner interface like in the case of [Map and Map.Entry<K,V>](https://www.programcreek.com/2013/08/inner-interface-in-java/).

JaniruTEC replied:
@infeo and I agreed to remove them entirely. 
Resolved with 7975114ce13382ea4ed326886a864924d8f4dcdf

JaniruTEC committed:
Changed maskValue to maskingValue in all MaskValueEnums

See: https://github.com/dokan-dev/dokan-java/pull/41#discussion_r432084032

infeo commented:
> conv does indeed stand for conversion. While I understand your suggestion wrapper wouldn't be a good name either because it's already taken by the #32-API

Because the other PR is currently only in a draft state, we can act independetly of it from my point of view.  But even if we consider it, why is it a problem to add both in the same package? The MaskValueSet and others are also convenience classes for easier use.  And finally, nowadays package names can be easily changed due to the great IDE support.

JaniruTEC committed:
Renamed getSetFromInt and getMaskValue

Renamed getSetFromInt to maskValueSet (and renamed parameter value to mask)
See: https://github.com/dokan-dev/dokan-java/pull/41#discussion_r432088419

Renamed getMaskValue to getMaskingValue
See: https://github.com/dokan-dev/dokan-java/pull/41#discussion_r432095165

JaniruTEC committed:
Renamed getSetFromInt to maskValueSet (and changed parameter name)

See: https://github.com/dokan-dev/dokan-java/pull/41#discussion_r432088419

JaniruTEC committed:
Merge branch 'improved-enumintegers' of https://github.com/JaniruTEC/dokan-java into improved-enumintegers

JaniruTEC commented:
> Because the other PR is currently only in a draft state, we can act independetly of it from my point of view. But even if we consider it, why is it a problem to add both in the same package? The MaskValueSet and others are also convenience classes for easier use. And finally, nowadays package names can be easily changed due to the great IDE support.

Putting too many classes into a package can lead to even more confusion than misleading package names. Also the EnumInteger-API is pretty much decoupled from the other wrapper classes and used over the whole project. That's why I wanted to keep it separate.

JaniruTEC commented:
Also to have said it: 
e042b3d and 3114b10 are the same commit, but 3114b10 has a more detailed message. I somehow messed up with amending the commit. Apologies

JaniruTEC committed:
Renamed getMaskingValue to maskingValue in MaskValueEnum

Renamed getMaskingValue to maskingValue in MaskValueEnum to match "intValue" in EnumInteger

JaniruTEC committed:
Fixed MaskValueSetTest and updated it to reformed EnumInteger-API

JaniruTEC committed:
Fixed remaining Tests to use reformed EnumInteger-API

infeo commented:
> 
> 
> > Because the other PR is currently only in a draft state, we can act independetly of it from my point of view. But even if we consider it, why is it a problem to add both in the same package? The MaskValueSet and others are also convenience classes for easier use. And finally, nowadays package names can be easily changed due to the great IDE support.
> 
> the EnumInteger-API is pretty much decoupled from the other wrapper classes and used over the whole project. That's why I wanted to keep it separate.

Good point. How about renaming the package to `enumInteger`, the name the main interface?

JaniruTEC committed:
Renamed package conv to masking

JaniruTEC commented:
@infeo and I agreed on "masking" as package name.

JaniruTEC committed:
Added IntegerConvertible as base interface for EnumInteger and MaskValueSet

Added IntegerConvertible as base interface for EnumInteger and MaskValueSet. IntergerConvertible defines the intValue() method and marks classes that can be converted into ints and vice-versa.
Changed EnumInteger to explicitly wrap Enums as special case of IntegerConvertible
Changed MaskValueSet to extend IntegerConvertible and renamed method toInt to intValue in accordance to IntegerConvertible
Changed method calls to refelct this change

JaniruTEC committed:
Removed delegating methods from EnumInteger

See: https://github.com/dokan-dev/dokan-java/pull/41#discussion_r432771434
